fendermint/vm/interpreter/src/chain.rs

```diff
@@ -318,3 +413,23 @@ fn relayed_bottom_up_ckpt_to_fvm(
 
     Ok(msg)
 }
+
+/// Encode to fvm implicit message
+pub fn encode_to_fvm_implicit(bytes: &[u8]) -> anyhow::Result<FvmMessage> {
+    let params = RawBytes::serialize(BytesSer(bytes))?;
+    let msg = FvmMessage {
+        version: 0,
+        from: system::SYSTEM_ACTOR_ADDR,
+        to: ipc::GATEWAY_ACTOR_ADDR,
+        value: TokenAmount::zero(),
```

Comment:
@adlrocha can you walk me through what happens when somebody wants to `fund` their account by locking up tokens in the parent subnet and sending them to the child? 

I'd expect that somewhere we need to parse messages and recognise which ones are top-down messages (to=gateway, method=apply...), parse the cross messages, find the `fund` types, sum up their values and add them to the `circ_supply`. 

A similar thing should happen with bottom-up messages: `release` should lower the `circ_supply`. 

These are missing at the moment. 

Nevertheless, the message goes to the Gateway who is supposed to send the amount to the recipient Ethereum account. But can the Gateway _mint_ tokens, or can it only send what it has? Should perhaps this `value` here be used as a minting mechanism, so that incoming messages increase the balance of the `Gateway` itself, in order for it to be able to credit it to the funded accounts? Can the system account mint tokens?

Reply:
1. When a user wants to fund an account in a subnet it calls [`fund` in the gateway of the parent](https://github.com/consensus-shipyard/ipc-solidity-actors/blob/e2624a563072a33b96ede3d3aafe0d78f87ed4ce/src/gateway/GatewayManagerFacet.sol#L130)
2. This function commits the [top-down message](https://github.com/consensus-shipyard/ipc-solidity-actors/blob/e2624a563072a33b96ede3d3aafe0d78f87ed4ce/src/gateway/GatewayManagerFacet.sol#L140), and stores the message for propagation in [the top down queue, updating the circulating supply, nonces, etc.](https://github.com/consensus-shipyard/ipc-solidity-actors/blob/63b3bf1f5f2774f31b25e4894982925d5a7a4672/src/lib/LibGateway.sol#L161-L165).

Fund is always top-down and release bottom-up, but when an arbitrary message arrives to the actor [`commitCrossMessage` is the one responsible for routing it as a top-down or a bottom-up](https://github.com/consensus-shipyard/ipc-solidity-actors/blob/63b3bf1f5f2774f31b25e4894982925d5a7a4672/src/gateway/GatewayMessengerFacet.sol#L76).

> But can the Gateway mint tokens, or can it only send what it has? Should perhaps this value here be used as a minting mechanism, so that incoming messages increase the balance of the Gateway itself, in order for it to be able to credit it to the funded accounts? Can the system account mint tokens?

This is a great question, what we did in Lotus is pre-allocate the maximum circulating supply that we want to allow in the subnet to the gateway, so the circulating supply is pre-minted and made available to the gateway to mind an burn when needed (actually, when we say burn what we should actually do is send the balance back to the gateway to avoid reducing the circulating supply).

That being said, now that we don't have the limitations from Lotus, I agree that it would be best to have ad-hoc minting of new funds here. It is a more elegant solution, and it prevents all of the pre-allocation and burning dance that we currently have.